wallet_migrator/src/lib.rs

```diff
@@ -0,0 +1,67 @@
+pub mod credx2anoncreds;
+pub mod error;
+pub mod vdrtools2credx;
+
+use std::fmt::Display;
+
+use error::MigrationResult;
+use log::{debug, error, info, warn};
+pub use vdrtools::{
+    types::domain::wallet::{Config, Credentials, KeyDerivationMethod, Record},
+    Locator, WalletHandle,
+};
+
+/// Retrieves all records from a wallet and migrates them
+/// by applying the `migrate_fn` argument.
+///
+/// The migrated records are inserted into a newly created
+/// wallet, based on the provided `config` and `credentials`.
+pub async fn migrate_wallet<E>(
+    wallet_handle: WalletHandle,
+    config: Config,
+    credentials: Credentials,
+    migrate_fn: impl FnMut(Record) -> Result<Option<Record>, E>,
+) -> MigrationResult<()>
+where
+    E: Display,
+{
+    info!("Starting wallet migration...");
+    let locator = Locator::instance();
+
+    debug!("Creating new wallet {} ...", config.id);
+    locator
+        .wallet_controller
+        .create(config.clone(), credentials.clone())
+        .await?;
+
+    debug!("Opening new wallet...");
```

Comment:
Assuming you agree on https://github.com/hyperledger/aries-vcx/pull/867/files#r1248167895
and we would work in this function with `source_config` and `target_config`, I'd tweak logs to include respective "config" when referring to one of the wallets. For example, here I would have logs like:

```
info!("Opening source wallet {:?}", source_config)
info!("Opening target wallet {:?}", target_config)
```

Also I think these high level processing logs deserve `info` rather than `debug`

Reply:
The configs contain some keys and other stuff that's not important outside of a debugging situation. 

You really only care about what happens in the info logs, with the minimal identification data (like the wallet.name).

But I will include the credentials and config, perhaps with some secret values.

Reply:
Ahha I see, I missed that config can contain secret values actually, good catch. Then I agree, lets just log wallet name (rather the wallet handle)

Reply:
There are also `trace` logs in `libvdrtools` with the full `credentials` and `config`, so maybe we can also leave this as it is, with just the wallet name getting logged as `debug`.

Reply:
When I run the migration, I would like to have easy way to see migrator's logs (without having to fiddle with rust_log and setup logs for migrator crate specifically)
I would like to see:
- when migrations process starts (info)
- creation of wallet - start and finish - as info
- want to see it's progressing (on per record that could be debug or trace, but you didn't address that), ideally perhaps every 100th migration record also info log
- when migration finish (info)
- deletion of wallet (on error) as warn

Also in general:
- be bit more explicit in logs, eg:
```
debug!("Opening new wallet...");
```
would be better to have something like:
```
debug!("Opening migration target wallet ${config.name}");
```

Additionally, I can't help it but:
```
debug!("Migrating records from wallet with handle {wallet_handle:?} to wallet with handle {new_wh:?}");
```
I find the inability to produce meaningful log (wallet handle its are not) message in **migrator** about 2 most important things - source and destination - are codesmell related to the signature. This kinda brings me again thoughts that migrator should be responsible for opening both wallets, and should be aware of what the source and destination wallets are.

I find logging pretty important for tool for this kind of operational tool like migrator. If something goes wrong, you really want to know what happened.

Reply:
Most of the stuff was already there...

1) when migrations process starts (info) - definitely was already there.
2) creation of wallet - start and finish - as info - the creation was there, though as `debug`. I assumed the finish is implied.
3) want to see it's progressing (on per record that could be debug or trace, but you didn't address that), ideally perhaps every 100th migration record also info log -> This is done in the `migrate_fn`, the one provided as `vdrtools2credx::migrate_any_records` already logs the initial and migrated records as `trace`. And, by all means, an arbitrary closure can be passed that wraps this function, or per type ones in `conv.rs`
4) when migration finish (info) - wasn't there, but I again assumed this is implied.
5) deletion of wallet (on error) as warn - was there, but as `error`.

Personally I don't see this as a "tool" (it's not like some CLI), but a library. And it's flexible enough that you can do whatever you want on a record by record basis.

In any case, I've been working on adding logging and change the signature to use two `Config` and `Credentials` sets when I decided to look into `WalletHandle` some more. I don't know why I was so convinced that multiple wallet instances can be open with different handles, but it turns out they can't. Opening the same wallet a second time fails. This is my bad, sorry.

This essentially means that accepting two handles and checking for equality is sufficient to ensure that the source and destination wallets are not the same, so I implemented that. This should solve the asymmetry concern and make most of the wallet logging obsolete.

Reply:
Maybe on short, the function now takes two `WalletHandle`s, errors out if they are the same, and the record logging was already there at a `trace` level. Both the old and the migrated records get logged.

Please let me know if this is not sufficient.